vnext/Shared/Networking/OriginPolicyHttpFilter.cpp

```diff
@@ -0,0 +1,638 @@
+  // CORS preflight should always exclude credentials although the subsequent CORS request may include credentials.
+  // if (!CheckAccessStatic(allowedOrigin, allowCredentials, GetOrigin(), withCredentials, /*out*/ errorText))
+  ValidateAllowOrigin(controlValues.AllowedOrigin, controlValues.AllowedCredentials, iRequestArgs);
+
+  // Check if the request method is allowed
+  // if (!IsCrossOriginRequestMethodAllowed(requestMethod, allowedMethodsList, withCredentials, /*out*/ errorText))
+  bool withCredentials = winrt::get_self<RequestArgs, IInspectable>(iRequestArgs)->WithCredentials;
+  bool requestMethodAllowed = false;
+  for (const auto &method : controlValues.AllowedMethods) {
+    if (L"*" == method) {
+      if (!withCredentials) {
+        requestMethodAllowed = true;
+        break;
+      }
+    } else if (boost::iequals(method, request.Method().ToString())) {
+      requestMethodAllowed = true;
+      break;
+    }
+  }
+  if (!requestMethodAllowed)
+    throw hresult_error{
+        E_INVALIDARG,
+        L"Method [" + request.Method().ToString() +
+            L"] is not allowed by Access-Control-Allow-Methods in preflight response.\\n"};
+
+  // Check if request headers are allowed
+  // if (!IsCrossOriginRequestHeadersAllowed(requestHeaders, allowedHeadersList, withCredentials, /*out*/ errorText))
+  // See https://fetch.spec.whatwg.org/#cors-preflight-fetch, section 4.8.7.6-7
+  // Check if the header should be allowed through wildcard, if the request does not have credentials.
+  bool requestHeadersAllowed = false;
+  if (!withCredentials && controlValues.AllowedHeaders.find(L"*") != controlValues.AllowedHeaders.cend()) {
+    // "Authorization" header cannot be allowed through wildcard alone.
+    // "Authorization" is the only member of https://fetch.spec.whatwg.org/#cors-non-wildcard-request-header-name.
+    if (request.Headers().HasKey(L"Authorization") && controlValues.AllowedHeaders.find(L"Authorization") == controlValues.AllowedHeaders.cend())
+      throw hresult_error{
+          E_INVALIDARG,
+          L"Request header field [Authorization] is not allowed by Access-Control-Allow-Headers in preflight response.\\n"};
+
+    requestHeadersAllowed = true;
+  }
+  if (!requestHeadersAllowed) {
+    // Forbidden headers are excluded from the JavaScript layer.
+    // User agents may use these headers internally.
+    // TODO: CorsUnsafeNotForbiddenRequestHeaderNames(requestHeaders);
+    const set unsafeNotForbidenHeaderNames = CorsUnsafeNotForbiddenRequestHeaderNames(request.Headers());
+    for (const auto name : unsafeNotForbidenHeaderNames) {
+      if (controlValues.AllowedHeaders.find(name) == controlValues.AllowedHeaders.cend())
+        throw hresult_error{
+            E_INVALIDARG,
+            L"Request header field [" + to_hstring(name) +
+                L"]is not allowed by Access-Control-Allow-Headers in preflight response.\\n"};
```

Comment:
>\\ [](http://example.com/codeflow?start=88&length=2)

(nit) Just curious, why is this \\n instead of \n?

Reply:
That string is ultimately returned to the JS layer, so you want the actual `\n`.
That said, I'll be removing those. We shouldn't add trailing newlines because we don't do it in any other native module.